Migrate all usages of Constraint as costs to use Cost class explicitly. Enable dynamic dispatch for adding generic Constraints / Costs.

EricCousineau-TRI opened the pull request:
This implements Step 2 of the PR Plan for Issue #5890.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/5950)
<!-- Reviewable:end -->

EricCousineau-TRI commented:
+@hongkai-dai for feature review
+@soonho-tri for platform review

---

Review status: 0 of 20 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-:-KifYF-XPNQ5Hp7Ckknt:b-kds9p0)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Reviewed 3 of 20 files at r1.
Review status: 3 of 20 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/solvers/constraint.h, line 22 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-Kih_myeempnUoV73fzL:-Kih_myeempnUoV73fzM:b-1omm3v) ([raw file](https://github.com/robotlocomotion/drake/blob/7fa43a5ac513300dc7dc7e299e0c330afc1e279e/drake/solvers/constraint.h#L22)):*
> ```objc
> namespace solvers {
> 
> class EvaluatorBase {
> ```

Should this be an abstract class?

---

*[drake/solvers/cost.h, line 23 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-KihXMa5YVa_vo1l56-s:-KihXMa6OWuWlGJ3i1qY:bbgbzoh) ([raw file](https://github.com/robotlocomotion/drake/blob/7fa43a5ac513300dc7dc7e299e0c330afc1e279e/drake/solvers/cost.h#L23)):*
> ```objc
>  * from cost, thus be convertible to a cost.
>  */
> class Cost : public EvaluatorBase {
> ```

Cost does not have lower or upper bound. How about removing the lower and upper bound from `EvaluatorBase`, and move to `Constraint`.

---

*[drake/solvers/cost.h, line 30 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-KihYc7Xkid0veBqCyDF:-KihYc7Xkid0veBqCyDG:b-thldn0) ([raw file](https://github.com/robotlocomotion/drake/blob/7fa43a5ac513300dc7dc7e299e0c330afc1e279e/drake/solvers/cost.h#L30)):*
> ```objc
> };
> 
> class CostShimBase : public Cost {
> ```

Do you expect to instantiate a concrete `CostShimBase` object? If not, maybe it makes sense to make it an abstract class?

---

*[drake/solvers/cost.h, line 46 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-KihYykYklenKgLZIu6s:-KihYykYklenKgLZIu6t:boe1h8y) ([raw file](https://github.com/robotlocomotion/drake/blob/7fa43a5ac513300dc7dc7e299e0c330afc1e279e/drake/solvers/cost.h#L46)):*
> ```objc
>               AutoDiffVecXd& y) const override;
> 
>   std::shared_ptr<Constraint> impl_;
> ```

data member should be private, according to google style guide. https://google.github.io/styleguide/cppguide.html#Access_Control

---

*[drake/solvers/cost.h, line 66 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-KihZeIXTf01Vq9B_-HX:-KihZeIXTf01Vq9B_-HY:b86j9bk) ([raw file](https://github.com/robotlocomotion/drake/blob/7fa43a5ac513300dc7dc7e299e0c330afc1e279e/drake/solvers/cost.h#L66)):*
> ```objc
> 
>  protected:
>   std::shared_ptr<C> constraint_;
> ```

data member should be private.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5950)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 3 of 20 files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/solvers/create_cost.h, line 50 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-KihdOJG1DIcI1SZLa-j:-KihdOJG1DIcI1SZLa-k:bkqo5td) ([raw file](https://github.com/robotlocomotion/drake/blob/7fa43a5ac513300dc7dc7e299e0c330afc1e279e/drake/solvers/create_cost.h#L50)):*
> ```objc
>  * Enable us to catch and provide a meaningful assertion if a Constraint is
>  * passed in, when we should have a Cost.
>  * @tparam C
> ```

Document `tparam C` and `tparam `F`?

---

*[drake/solvers/create_cost.h, line 81 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-KihdmBh5tZl43dq2ZYj:-KihdmBh5tZl43dq2ZYk:bw4fuzl) ([raw file](https://github.com/robotlocomotion/drake/blob/7fa43a5ac513300dc7dc7e299e0c330afc1e279e/drake/solvers/create_cost.h#L81)):*
> ```objc
>   // Use deferred evaluation
>   static_assert(!value, "You cannot pass a Constraint to "
>       "create a FunctionCost object. Please ensure you are passing a Cost.");
> ```

This message only makes sense if the user wants to call `assert_if_is_constraint` when creating the constraint. If the user uses this struct in some other context, just to check if the type `F` is a constraint, then this message does not make sense.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5950)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Review status: 1 of 20 files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/solvers/constraint.h, line 22 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-Kih_myeempnUoV73fzL:-KikQdgWpaNiNRIWzHY4:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/7fa43a5ac513300dc7dc7e299e0c330afc1e279e/drake/solvers/constraint.h#L22)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Should this be an abstract class?
</blockquote></details>

Done.

---

*[drake/solvers/cost.h, line 23 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-KihXMa5YVa_vo1l56-s:-KijtS9vslkSPfKNlLoW:bp6pr7k) ([raw file](https://github.com/robotlocomotion/drake/blob/7fa43a5ac513300dc7dc7e299e0c330afc1e279e/drake/solvers/cost.h#L23)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Cost does not have lower or upper bound. How about removing the lower and upper bound from `EvaluatorBase`, and move to `Constraint`.
</blockquote></details>

I'd prefer to defer changing the `EvaluatorBase` for another issue (and generalizing other things, such as changing `Binding::constraint()`, etc.). Would you be OK if I created a separate issue for this?

---

*[drake/solvers/cost.h, line 30 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-KihYc7Xkid0veBqCyDF:-KijxouEBuoGDLDaAyGP:bn2r1sr) ([raw file](https://github.com/robotlocomotion/drake/blob/7fa43a5ac513300dc7dc7e299e0c330afc1e279e/drake/solvers/cost.h#L30)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Do you expect to instantiate a concrete `CostShimBase` object? If not, maybe it makes sense to make it an abstract class?
</blockquote></details>

I have intended it as abstract given that it only provides a `protected` constructor and does not inherit constructors. Would you prefer to see another mechanism to make it abstract? (I could add a pure virtual method, but that would get close to invalidating the purpose of this class.)

---

*[drake/solvers/cost.h, line 46 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-KihYykYklenKgLZIu6s:-KijylRGN8s61you2oOh:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/7fa43a5ac513300dc7dc7e299e0c330afc1e279e/drake/solvers/cost.h#L46)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

data member should be private, according to google style guide. https://google.github.io/styleguide/cppguide.html#Access_Control
</blockquote></details>

Done.

---

*[drake/solvers/cost.h, line 66 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-KihZeIXTf01Vq9B_-HX:-KijztKRVS8h9MzkOO3K:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/7fa43a5ac513300dc7dc7e299e0c330afc1e279e/drake/solvers/cost.h#L66)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

data member should be private.
</blockquote></details>

Done.

---

*[drake/solvers/create_cost.h, line 50 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-KihdOJG1DIcI1SZLa-j:-Kik7_ZDWb0gy39W3Vk9:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/7fa43a5ac513300dc7dc7e299e0c330afc1e279e/drake/solvers/create_cost.h#L50)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Document `tparam C` and `tparam `F`?
</blockquote></details>

Done.

---

*[drake/solvers/create_cost.h, line 81 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-KihdmBh5tZl43dq2ZYj:-Kik-HilvnN-5MO1xySc:b-voz7ou) ([raw file](https://github.com/robotlocomotion/drake/blob/7fa43a5ac513300dc7dc7e299e0c330afc1e279e/drake/solvers/create_cost.h#L81)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

This message only makes sense if the user wants to call `assert_if_is_constraint` when creating the constraint. If the user uses this struct in some other context, just to check if the type `F` is a constraint, then this message does not make sense.
</blockquote></details>

The purpose of this is to implement the `AssertionChecker` template pattern:
http://stackoverflow.com/a/13366183/7829525

It is explicitly meant to catch any `Constraint` that was inadvertently passed to this function, so that the user does not have to schlep through a slew of template binding reports.

Usage in this PR:
https://github.com/RobotLocomotion/drake/pull/5950/files#diff-2594a593d42a6e8d4baa0b37959def7aR692
(Trying to figure out how to link to lines in Reviewable... Also, you may have not seen it earlier, as Reviewable had collapsed the revisions to diff, and I forgot to expand them.)

Otherwise, if a user wishes to detect if an argument is `Constraint`-compatible, they could use `is_binding_compatible<F, Constraint>::value` directly.

(Added documentation)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5950)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 14 of 20 files at r1, 4 of 4 files at r2.
Review status: 19 of 20 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/solvers/constraint.h, line 155 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-Kil92IOrK759w7Yx4OF:-Kil92IOrK759w7Yx4OG:bxcj6m9) ([raw file](https://github.com/robotlocomotion/drake/blob/71f1e6de5d9e57fea1c67bcadb1dfb62122ce764/drake/solvers/constraint.h#L155)):*
> ```objc
> 
>  protected:
>   /// Constructs a constraint which has \p num_constraints rows, with input
> ```

BTW, you can use single quotation mark around num_constraint, doxygen will also render that. It is probably easier to read than `\p num_constraint`.

---

*[drake/solvers/cost.h, line 23 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-KihXMa5YVa_vo1l56-s:-Kil8BjrE9JP4Nj_3jYe:b-ysqs5n) ([raw file](https://github.com/robotlocomotion/drake/blob/7fa43a5ac513300dc7dc7e299e0c330afc1e279e/drake/solvers/cost.h#L23)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

I'd prefer to defer changing the `EvaluatorBase` for another issue (and generalizing other things, such as changing `Binding::constraint()`, etc.). Would you be OK if I created a separate issue for this?
</blockquote></details>

A separate issue sounds good.

---

*[drake/solvers/cost.h, line 30 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-KihYc7Xkid0veBqCyDF:-Kil8I7bgA9XFUlnPmTG:b-v7jekp) ([raw file](https://github.com/robotlocomotion/drake/blob/7fa43a5ac513300dc7dc7e299e0c330afc1e279e/drake/solvers/cost.h#L30)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

I have intended it as abstract given that it only provides a `protected` constructor and does not inherit constructors. Would you prefer to see another mechanism to make it abstract? (I could add a pure virtual method, but that would get close to invalidating the purpose of this class.)
</blockquote></details>

Sorry, I did not catch the protected part. Makes sense.

---

*[drake/solvers/create_cost.h, line 81 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-KihdmBh5tZl43dq2ZYj:-KilA5vb4XtuBBpxQSby:b6c9vfd) ([raw file](https://github.com/robotlocomotion/drake/blob/7fa43a5ac513300dc7dc7e299e0c330afc1e279e/drake/solvers/create_cost.h#L81)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

The purpose of this is to implement the `AssertionChecker` template pattern:
http://stackoverflow.com/a/13366183/7829525

It is explicitly meant to catch any `Constraint` that was inadvertently passed to this function, so that the user does not have to schlep through a slew of template binding reports.

Usage in this PR:
https://github.com/RobotLocomotion/drake/pull/5950/files#diff-2594a593d42a6e8d4baa0b37959def7aR692
(Trying to figure out how to link to lines in Reviewable... Also, you may have not seen it earlier, as Reviewable had collapsed the revisions to diff, and I forgot to expand them.)

Otherwise, if a user wishes to detect if an argument is `Constraint`-compatible, they could use `is_binding_compatible<F, Constraint>::value` directly.

(Added documentation)
</blockquote></details>

Thanks for the explanation, that is very clear.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-:-KilCLnlS7PQYPw8NFat:bnfp4nl)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Review status: 19 of 20 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/solvers/constraint.h, line 155 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-Kil92IOrK759w7Yx4OF:-KiogKDGlZZCGp4PaA-r:bo5p0qg) ([raw file](https://github.com/robotlocomotion/drake/blob/71f1e6de5d9e57fea1c67bcadb1dfb62122ce764/drake/solvers/constraint.h#L155)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

BTW, you can use single quotation mark around num_constraint, doxygen will also render that. It is probably easier to read than `\p num_constraint`.
</blockquote></details>

Nice, didn't know that. Thanks!

---

*[drake/solvers/create_cost.h, line 81 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-KihdmBh5tZl43dq2ZYj:-KiogU3Cg76SMDB0jjpb:b-8p8n0v) ([raw file](https://github.com/robotlocomotion/drake/blob/7fa43a5ac513300dc7dc7e299e0c330afc1e279e/drake/solvers/create_cost.h#L81)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Thanks for the explanation, that is very clear.
</blockquote></details>

Welcome!
And would you be able to mark this as Satisfied? I believe Reviewable still believes it needs review.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5950)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Reviewed 1 of 20 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5950)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Check point. 

---

Reviewed 2 of 20 files at r1, 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 15 unresolved discussions.

---

*[drake/solvers/constraint.h, line 22 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-Kior4btXZ5Elea7uCPV:-Kior4btXZ5Elea7uCPW:b6t4qax) ([raw file](https://github.com/robotlocomotion/drake/blob/71f1e6de5d9e57fea1c67bcadb1dfb62122ce764/drake/solvers/constraint.h#L22)):*
> ```objc
> namespace solvers {
> 
> class EvaluatorBase {
> ```

Missing doxygen documentation for the class `EvaluateBase`.

---

*[drake/solvers/constraint.h, line 23 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-KiorGU6JfXRCcpRAexz:-KiorGU6JfXRCcpRAey-:bu39w00) ([raw file](https://github.com/robotlocomotion/drake/blob/71f1e6de5d9e57fea1c67bcadb1dfb62122ce764/drake/solvers/constraint.h#L23)):*
> ```objc
> 
> class EvaluatorBase {
>   void check(size_t num_constraints) {
> ```

This `private` method needs to be moved to the `private:` section at the end. I understand that it's not you who put this one here, but let's clean things up while we are working on refactoring :-)

---

*[drake/solvers/constraint.h, line 36 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-KiorfSJHRE1gfobK2gu:-KiorfSJHRE1gfobK2gv:b-952e9r) ([raw file](https://github.com/robotlocomotion/drake/blob/71f1e6de5d9e57fea1c67bcadb1dfb62122ce764/drake/solvers/constraint.h#L36)):*
> ```objc
>   template <typename DerivedLB, typename DerivedUB>
>   EvaluatorBase(size_t num_constraints, int num_vars,
>                 Eigen::MatrixBase<DerivedLB> const& lb,
> ```

META: `T const&` -> `const T&`

---

*[drake/solvers/cost.h, line 30 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-Kiot0HDzISWIswdV2KQ:-Kiot0HDzISWIswdV2KR:bnjwd3l) ([raw file](https://github.com/robotlocomotion/drake/blob/71f1e6de5d9e57fea1c67bcadb1dfb62122ce764/drake/solvers/cost.h#L30)):*
> ```objc
> };
> 
> class CostShimBase : public Cost {
> ```

Missing doxygen documentation for the class `CostShimBase`. Also please check the documentation of `Cost` class above.

---

*[drake/solvers/cost.h, line 49 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-KiouOTCto03GhgxES7-:-KiouOTCto03GhgxES70:boizw5) ([raw file](https://github.com/robotlocomotion/drake/blob/71f1e6de5d9e57fea1c67bcadb1dfb62122ce764/drake/solvers/cost.h#L49)):*
> ```objc
> 
>  private:
>   std::shared_ptr<Constraint> impl_;
> ```

Better to make it `const`? (unless we have a plan to update it in the future PRs). 

---

*[drake/solvers/cost.h, line 65 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-KiougCYvckcneltlqK9:-KiougCYvckcneltlqKA:b-mhiyiv) ([raw file](https://github.com/robotlocomotion/drake/blob/71f1e6de5d9e57fea1c67bcadb1dfb62122ce764/drake/solvers/cost.h#L65)):*
> ```objc
>   explicit CostShim(Args&&... args)
>       : CostShimBase(std::make_shared<C>(std::forward<Args>(args)...)) {
>     constraint_ = std::dynamic_pointer_cast<C>(impl());
> ```

Do we need to worry about the case where `std::dynamic_pointer_cast` fails? If yes, let's do some error-handling (i.e. `throw`). If no, let's use `static_pointer_cast` and add a comment justifying the invariant.

---

*[drake/solvers/cost.h, line 82 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-KiovCUr1vyACiNahO3U:-KiovCUr1vyACiNahO3V:bbnezes) ([raw file](https://github.com/robotlocomotion/drake/blob/71f1e6de5d9e57fea1c67bcadb1dfb62122ce764/drake/solvers/cost.h#L82)):*
> ```objc
>   DRAKE_NO_COPY_NO_MOVE_NO_ASSIGN(LinearCost)
> 
>   // NOLINTNEXTLINE(runtime/explicit).
> ```

FYI, this comment is not necessary.

Instead, we need to mark this constructor as `explicit`.

---

*[drake/solvers/cost.h, line 115 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-KiovunkLNNEpT9nBUVZ:-KiovunkLNNEpT9nBUV_:b-65n1jj) ([raw file](https://github.com/robotlocomotion/drake/blob/71f1e6de5d9e57fea1c67bcadb1dfb62122ce764/drake/solvers/cost.h#L115)):*
> ```objc
>    * Updates the quadratic and linear term of the constraint. The new
>    * matrices need to have the same dimension as before.
>    * @param new_Q new quadratic term
> ```

BTW, missing periods at the end. (this one and the below)

---

*[drake/solvers/create_cost.h, line 48 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-KiowjkECXGPlxflJjTe:-KiowjkECXGPlxflJjTf:bnr5t5w) ([raw file](https://github.com/robotlocomotion/drake/blob/71f1e6de5d9e57fea1c67bcadb1dfb62122ce764/drake/solvers/create_cost.h#L48)):*
> ```objc
> 
> /**
>  * Enable us to catch and provide a meaningful assertion if a Constraint is
> ```

nit: `Enable` -> `Enables`.

---

*[drake/solvers/mathematical_program.cc, line 434 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-Kip0GiGcETDaK06Hwiu:-Kip0GiGcETDaK06Hwiv:b3ge3vn) ([raw file](https://github.com/robotlocomotion/drake/blob/71f1e6de5d9e57fea1c67bcadb1dfb62122ce764/drake/solvers/mathematical_program.cc#L434)):*
> ```cpp
> template <typename To, typename From>
> Binding<To> BindingUpcast(const Binding<From>& binding) {
>   return Binding<To>(std::dynamic_pointer_cast<To>(binding.constraint()),
> ```

It seems that it's safe to use `static_pointer_cast` here because all call-sites have the check? If so, let's switch to `static_pointer_cast` and document the pre-condition. Having a `DRAKE_ASSERT` is also a good idea.

---

*[drake/solvers/mathematical_program.h, line 656 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-KiozL2D7LPxXqydckHN:-KiozL2D7LPxXqydckHO:b5cpigs) ([raw file](https://github.com/robotlocomotion/drake/blob/71f1e6de5d9e57fea1c67bcadb1dfb62122ce764/drake/solvers/mathematical_program.h#L656)):*
> ```objc
> 
>   /**
>    * Convert an input of type @tparam F to a FunctionCost object.
> ```

BTW, `@tparam` -> `@p`.

---

*[drake/solvers/mathematical_program.h, line 666 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-KiozbiqrF7a1DxpxCFO:-KiozbiqrF7a1DxpxCFP:b-18b437) ([raw file](https://github.com/robotlocomotion/drake/blob/71f1e6de5d9e57fea1c67bcadb1dfb62122ce764/drake/solvers/mathematical_program.h#L666)):*
> ```objc
> 
>   /**
>    * Add costs to the optimization program on a list of variables.
> ```

BTW, `Add costs` -> `Adds a cost`?

---

*[drake/solvers/mathematical_program.h, line 699 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-Kip-JMqu7GQVQ7cmz1p:-Kip-JMqu7GQVQ7cmz1q:b6es5xt) ([raw file](https://github.com/robotlocomotion/drake/blob/71f1e6de5d9e57fea1c67bcadb1dfb62122ce764/drake/solvers/mathematical_program.h#L699)):*
> ```objc
>   typename std::enable_if<detail::assert_if_is_constraint<F>::value,
>                           Binding<Cost>>::type
>   AddCost(F&& f, Vars&& vars) {
> ```

BTW, ->
```c++
 AddCost(F&&, Vars&&)
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-:-Kip1MTwcDm7BdAUNZcj:bgbo55x)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
First round completed. PTAL.

---

Reviewed 14 of 20 files at r1.
Review status: all files reviewed at latest revision, 18 unresolved discussions.

---

*[drake/bindings/pybind11/pydrake_mathematicalprogram.cc, line 16 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-Kip2beqXlO3QucYwKf8:-Kip2beqXlO3QucYwKf9:b-gbwrzl) ([raw file](https://github.com/robotlocomotion/drake/blob/71f1e6de5d9e57fea1c67bcadb1dfb62122ce764/drake/bindings/pybind11/pydrake_mathematicalprogram.cc#L16)):*
> ```cpp
> template <typename C>
> // NOLINTNEXTLINE(runtime/references).
> auto RegisterBinding(py::module &m, const string& name) {
> ```

Let's remove `// NOLINTNEXTLINE(runtime/references).` by using `py::module* m` instead of `py::module &m`. It seems that it only requires editing <=10 lines in this file.

---

*[drake/solvers/test/convex_optimization_test.cc, line 242 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-Kip3jXStfzvz776_0tr:-Kip3jXStfzvz776_0ts:b-yl5if3) ([raw file](https://github.com/robotlocomotion/drake/blob/71f1e6de5d9e57fea1c67bcadb1dfb62122ce764/drake/solvers/test/convex_optimization_test.cc#L242)):*
> ```cpp
>   prog_socp.AddLinearConstraint(A, b_lb, b_ub, x_socp);
> 
>   std::shared_ptr<LinearCost> cost_socp1(new LinearCost(c.transpose()));
> ```

BTW, better to rewrite it without `new` such as
```c++
  auto cost_socp1 = make_shared<LinearCost>(c.transpose());
```

---

*[drake/systems/trajectory_optimization/direct_collocation.cc, line 59 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-Kip5qq0E2GXE8rt1qHr:-Kip5qq0E2GXE8rt1qHs:bohziv6) ([raw file](https://github.com/robotlocomotion/drake/blob/71f1e6de5d9e57fea1c67bcadb1dfb62122ce764/drake/systems/trajectory_optimization/direct_collocation.cc#L59)):*
> ```cpp
> class RunningCostEndWrapper : public solvers::Cost {
>  public:
>   explicit RunningCostEndWrapper(std::shared_ptr<solvers::Cost> cost)
> ```

BTW, let's make it 
```c++
explicit RunningCostEndWrapper(const std::shared_ptr<solvers::Cost>& cost)
```

---

*[drake/systems/trajectory_optimization/direct_collocation.cc, line 76 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-Kip5vVsB7An7_RMLcYL:-Kip5vVsB7An7_RMLcYM:b-uhvoxj) ([raw file](https://github.com/robotlocomotion/drake/blob/71f1e6de5d9e57fea1c67bcadb1dfb62122ce764/drake/systems/trajectory_optimization/direct_collocation.cc#L76)):*
> ```cpp
> 
>  private:
>   std::shared_ptr<Cost> cost_;
> ```

and
```
const std::shared_ptr<Cost> cost_;
```

Same things for `RunningCostMidWrapper` class. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-:-Kip9K5svsT4a2Bgn0ww:b4bv8rt)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Review status: 12 of 20 files reviewed at latest revision, 17 unresolved discussions.

---

*[drake/solvers/constraint.h, line 22 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-Kior4btXZ5Elea7uCPV:-KipcUeVFtmuZmrN1Xqs:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/71f1e6de5d9e57fea1c67bcadb1dfb62122ce764/drake/solvers/constraint.h#L22)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

Missing doxygen documentation for the class `EvaluateBase`.
</blockquote></details>

Done.

---

*[drake/solvers/constraint.h, line 23 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-KiorGU6JfXRCcpRAexz:-KipcVJYTZcsbwRtAfLc:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/71f1e6de5d9e57fea1c67bcadb1dfb62122ce764/drake/solvers/constraint.h#L23)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

This `private` method needs to be moved to the `private:` section at the end. I understand that it's not you who put this one here, but let's clean things up while we are working on refactoring :-)
</blockquote></details>

Done.

---

*[drake/solvers/constraint.h, line 36 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-KiorfSJHRE1gfobK2gu:-KipcjCig59iU8BZmFQQ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/71f1e6de5d9e57fea1c67bcadb1dfb62122ce764/drake/solvers/constraint.h#L36)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

META: `T const&` -> `const T&`
</blockquote></details>

Done.

---

*[drake/solvers/cost.h, line 30 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-Kiot0HDzISWIswdV2KQ:-KipdC0JuZdXsMzWKPRY:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/71f1e6de5d9e57fea1c67bcadb1dfb62122ce764/drake/solvers/cost.h#L30)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

Missing doxygen documentation for the class `CostShimBase`. Also please check the documentation of `Cost` class above.
</blockquote></details>

Done.

---

*[drake/solvers/cost.h, line 49 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-KiouOTCto03GhgxES7-:-KipdKeLvMgyl64J2i7L:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/71f1e6de5d9e57fea1c67bcadb1dfb62122ce764/drake/solvers/cost.h#L49)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

Better to make it `const`? (unless we have a plan to update it in the future PRs). 
</blockquote></details>

Done.

---

*[drake/solvers/cost.h, line 65 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-KiougCYvckcneltlqK9:-KipdmivXq8kKZK7kcsU:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/71f1e6de5d9e57fea1c67bcadb1dfb62122ce764/drake/solvers/cost.h#L65)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

Do we need to worry about the case where `std::dynamic_pointer_cast` fails? If yes, let's do some error-handling (i.e. `throw`). If no, let's use `static_pointer_cast` and add a comment justifying the invariant.
</blockquote></details>

Done.

---

*[drake/solvers/cost.h, line 82 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-KiovCUr1vyACiNahO3U:-KipeDqkh83qb0VIaBub:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/71f1e6de5d9e57fea1c67bcadb1dfb62122ce764/drake/solvers/cost.h#L82)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

FYI, this comment is not necessary.

Instead, we need to mark this constructor as `explicit`.
</blockquote></details>

Done.

---

*[drake/solvers/cost.h, line 115 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-KiovunkLNNEpT9nBUVZ:-KipeQToTh7hJxxLPQnf:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/71f1e6de5d9e57fea1c67bcadb1dfb62122ce764/drake/solvers/cost.h#L115)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, missing periods at the end. (this one and the below)
</blockquote></details>

Done.

---

*[drake/solvers/create_cost.h, line 48 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-KiowjkECXGPlxflJjTe:-KipeXOHi9VEoD65OsAg:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/71f1e6de5d9e57fea1c67bcadb1dfb62122ce764/drake/solvers/create_cost.h#L48)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

nit: `Enable` -> `Enables`.
</blockquote></details>

Done.

---

*[drake/solvers/mathematical_program.cc, line 434 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-Kip0GiGcETDaK06Hwiu:-KipfSa65TpunpeNXPoY:baench5) ([raw file](https://github.com/robotlocomotion/drake/blob/71f1e6de5d9e57fea1c67bcadb1dfb62122ce764/drake/solvers/mathematical_program.cc#L434)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

It seems that it's safe to use `static_pointer_cast` here because all call-sites have the check? If so, let's switch to `static_pointer_cast` and document the pre-condition. Having a `DRAKE_ASSERT` is also a good idea.
</blockquote></details>

I'd prefer to keep this as a `dynamic_pointer_cast`, as we should eventually discard the use of RTTI to dispatch to the appropriate container.

I've added a `DRAKE_DEMAND` to accommodate this.

---

*[drake/solvers/mathematical_program.h, line 656 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-KiozL2D7LPxXqydckHN:-KipegLBR2BLWjh4xiBk:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/71f1e6de5d9e57fea1c67bcadb1dfb62122ce764/drake/solvers/mathematical_program.h#L656)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, `@tparam` -> `@p`.
</blockquote></details>

Done.

---

*[drake/solvers/mathematical_program.h, line 666 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-KiozbiqrF7a1DxpxCFO:-KipekcOcYq-D9__geU0:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/71f1e6de5d9e57fea1c67bcadb1dfb62122ce764/drake/solvers/mathematical_program.h#L666)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, `Add costs` -> `Adds a cost`?
</blockquote></details>

Done.

---

*[drake/solvers/mathematical_program.h, line 699 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-Kip-JMqu7GQVQ7cmz1p:-Kipevjq5mErhConNulJ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/71f1e6de5d9e57fea1c67bcadb1dfb62122ce764/drake/solvers/mathematical_program.h#L699)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, ->
```c++
 AddCost(F&&, Vars&&)
```
</blockquote></details>

Done.

---

*[drake/solvers/test/convex_optimization_test.cc, line 242 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-Kip3jXStfzvz776_0tr:-Kipf6g94ClrQfPt-iT_:b-jxw2vg) ([raw file](https://github.com/robotlocomotion/drake/blob/71f1e6de5d9e57fea1c67bcadb1dfb62122ce764/drake/solvers/test/convex_optimization_test.cc#L242)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, better to rewrite it without `new` such as
```c++
  auto cost_socp1 = make_shared<LinearCost>(c.transpose());
```
</blockquote></details>

Done. Replaced with `CreateLinearConst` :)

---

*[drake/systems/trajectory_optimization/direct_collocation.cc, line 59 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-Kip5qq0E2GXE8rt1qHr:-KipfFaNdQrqmnmYsDOi:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/71f1e6de5d9e57fea1c67bcadb1dfb62122ce764/drake/systems/trajectory_optimization/direct_collocation.cc#L59)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, let's make it 
```c++
explicit RunningCostEndWrapper(const std::shared_ptr<solvers::Cost>& cost)
```
</blockquote></details>

Done.

---

*[drake/systems/trajectory_optimization/direct_collocation.cc, line 76 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-Kip5vVsB7An7_RMLcYL:-KipfODO5dqsP6YXmgfT:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/71f1e6de5d9e57fea1c67bcadb1dfb62122ce764/drake/systems/trajectory_optimization/direct_collocation.cc#L76)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

and
```
const std::shared_ptr<Cost> cost_;
```

Same things for `RunningCostMidWrapper` class. 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5950)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Reviewed 8 of 8 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5950)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5950#-:-KipywEu-UMOEwnbYKRa:bnfp4nl)*
<!-- Sent from Reviewable.io -->
